File: src/levels/L13-command/ts/command.ts
// Un pedido convertido en objeto: se puede guardar, encolar, ejecutar después… o deshacer.
class OrderCommand {
  constructor(private barista: Barista, private drink: string, readonly orderId: number) {}
  execute() { this.barista.prepare(this.drink) }
}

class FrontDesk {
  constructor(private barista: Barista, private queue: CommandQueue) {}

  // region: FrontDesk#submit
  submit(request: OrderRequest) {
    if (request.action === 'cancelar') return this.queue.cancel(request.orderId)
    this.queue.push(new OrderCommand(this.barista, request.drink, request.orderId))
  }
  // endregion
}

class CommandQueue {
  private pending: OrderCommand[] = []

  // Los comandos esperan su turno; deshacer es sacarlos de la cola antes de que corran.
  // region: CommandQueue#push
  push(command: OrderCommand) { this.pending.push(command) }
  cancel(orderId: number) { this.pending = this.pending.filter((c) => c.orderId !== orderId) }
  runNext() { this.pending.shift()?.execute() }
  // endregion
}

File: src/levels/L13-command/command.rb
# Un pedido convertido en objeto: se puede guardar, encolar, ejecutar después… o deshacer.
class OrderCommand
  attr_reader :order_id

  def initialize(barista, drink, order_id) = (@barista, @drink, @order_id = barista, drink, order_id)
  def execute = @barista.prepare(@drink)
end

class FrontDesk
  # region: FrontDesk#submit
  def submit(request)
    return @queue.cancel(request.order_id) if request.action == :cancelar
    @queue.push(OrderCommand.new(@barista, request.drink, request.order_id))
  end
  # endregion
end

class CommandQueue
  # Los comandos esperan su turno; deshacer es sacarlos de la cola antes de que corran.
  # region: CommandQueue#push
  def push(command) = @pending << command
  def cancel(order_id) = @pending.reject! { |c| c.order_id == order_id }
  def run_next = @pending.shift&.execute
  # endregion
end
